File: ts/tests/fs-read.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { freshStore, toBytes, fromBytes, rmTmpDir, fs } from './helpers.js';
import {
  GitStore,
  FS,
  FileType,
  FileNotFoundError,
  IsADirectoryError,
  NotADirectoryError,
} from '../src/index.js';
import * as path from 'node:path';

let store: GitStore;
let snap: FS;
let tmpDir: string;

beforeEach(async () => {
  const res = await freshStore();
  store = res.store;
  tmpDir = res.tmpDir;

  let f = await store.branches.get('main');
  f = await f.write('hello.txt', toBytes('Hello!'));
  f = await f.write('src/main.py', toBytes("print('hi')"));
  snap = await f.write('src/lib/util.py', toBytes('# util'));
});

afterEach(() => rmTmpDir(tmpDir));

describe('read', () => {
  it('reads a file', async () => {
    expect(fromBytes(await snap.read('hello.txt'))).toBe('Hello!');
  });

  it('reads nested file', async () => {
    expect(fromBytes(await snap.read('src/main.py'))).toBe("print('hi')");
  });

  it('missing file throws FileNotFoundError', async () => {
    await expect(snap.read('nope.txt')).rejects.toThrow(FileNotFoundError);
  });

  it('reading a directory throws IsADirectoryError', async () => {
    await expect(snap.read('src')).rejects.toThrow(IsADirectoryError);
  });
});

describe('readText', () => {
  it('reads text', async () => {
    expect(await snap.readText('hello.txt')).toBe('Hello!');
  });

  it('reads with encoding', async () => {
    // Write raw latin-1 bytes for 'café'
    const bytes = new Uint8Array([0x63, 0x61, 0x66, 0xe9]);
    const f2 = await snap.write('latin.txt', bytes);
    const text = await f2.readText('latin.txt', 'latin1');
    expect(text).toBe('café');
  });
});

describe('ls', () => {
  it('lists root', async () => {
    const items = (await snap.ls()).sort();
    expect(items).toEqual(['hello.txt', 'src']);
  });

  it('lists subdirectory', async () => {
    const items = (await snap.ls('src')).sort();
    expect(items).toEqual(['lib', 'main.py']);
  });

  it('ls on file throws NotADirectoryError', async () => {
    await expect(snap.ls('hello.txt')).rejects.toThrow(NotADirectoryError);
  });
});

describe('walk', () => {
  it('walks root', async () => {
    const entries = [];
    for await (const e of snap.walk()) entries.push(e);
    const [dirpath, dirs, files] = entries[0];
    expect(dirpath).toBe('');
    expect(dirs).toContain('src');
    expect(files.map((f) => f.name)).toContain('hello.txt');
  });

  it('walks subdirectory', async () => {
    const entries = [];
    for await (const e of snap.walk('src')) entries.push(e);
    const [dirpath, , files] = entries[0];
    expect(dirpath).toBe('src');
    expect(files.map((f) => f.name)).toContain('main.py');
  });

  it('walk on file throws NotADirectoryError', async () => {
    const entries = [];
    await expect(async () => {
      for await (const e of snap.walk('hello.txt')) entries.push(e);
    }).rejects.toThrow(NotADirectoryError);
  });
});

describe('exists', () => {
  it('file exists', async () => {
    expect(await snap.exists('hello.txt')).toBe(true);
  });

  it('directory exists', async () => {
    expect(await snap.exists('src')).toBe(true);
  });

  it('missing returns false', async () => {
    expect(await snap.exists('nope.txt')).toBe(false);
  });
});

describe('copyOut root', () => {
  it('copyOut creates files on disk', async () => {
    const outDir = path.join(tmpDir, 'export');
    await snap.copyOut('/', outDir);
    expect(fs.readFileSync(path.join(outDir, 'hello.txt'), 'utf-8')).toBe('Hello!');
    expect(fs.readFileSync(path.join(outDir, 'src/main.py'), 'utf-8')).toBe("print('hi')");
    expect(fs.readFileSync(path.join(outDir, 'src/lib/util.py'), 'utf-8')).toBe('# util');
  });

  it('copyOut empty repository', async () => {
    const { store: s2, tmpDir: td2 } = await freshStore();
    const emptyFs = await s2.branches.get('main');
    const outDir = path.join(td2, 'export');
    fs.mkdirSync(outDir, { recursive: true });
    await emptyFs.copyOut('/', outDir);
    expect(fs.existsSync(outDir)).toBe(true);
    rmTmpDir(td2);
  });

  it('copyOut into existing directory', async () => {
    const outDir = path.join(tmpDir, 'export');
    fs.mkdirSync(outDir, { recursive: true });
    fs.writeFileSync(path.join(outDir, 'hello.txt'), 'old');
    await snap.copyOut('/', outDir);
    expect(fs.readFileSync(path.join(outDir, 'hello.txt'), 'utf-8')).toBe('Hello!');
  });

  it('copyOut symlinks', async () => {
    const f2 = await snap.writeSymlink('link.txt', 'hello.txt');
    const outDir = path.join(tmpDir, 'export-sym');
    await f2.copyOut('/', outDir);
    const target = fs.readlinkSync(path.join(outDir, 'link.txt'));
    expect(target).toBe('hello.txt');
  });

  it('copyOut symlinks overwriting regular files', async () => {
    const f2 = await snap.writeSymlink('hello.txt', 'src/main.py');
    const outDir = path.join(tmpDir, 'export-overwrite');
    fs.mkdirSync(outDir, { recursive: true });
    fs.writeFileSync(path.join(outDir, 'hello.txt'), 'old');
    await f2.copyOut('/', outDir);
    const target = fs.readlinkSync(path.join(outDir, 'hello.txt'));
    expect(target).toBe('src/main.py');
  });
});

describe('fileType', () => {
  it('blob type', async () => {
    expect(await snap.fileType('hello.txt')).toBe(FileType.BLOB);
  });

  it('tree type', async () => {
    expect(await snap.fileType('src')).toBe(FileType.TREE);
  });

  it('nested file type', async () => {
    expect(await snap.fileType('src/main.py')).toBe(FileType.BLOB);
  });

  it('executable type', async () => {
    const f2 = await snap.write('run.sh', toBytes('#!/bin/sh'), { mode: FileType.EXECUTABLE });
    expect(await f2.fileType('run.sh')).toBe(FileType.EXECUTABLE);
  });

  it('symlink type', async () => {
    const f2 = await snap.writeSymlink('link.txt', 'hello.txt');
    expect(await f2.fileType('link.txt')).toBe(FileType.LINK);
  });

  it('missing throws FileNotFoundError', async () => {
    await expect(snap.fileType('nope.txt')).rejects.toThrow(FileNotFoundError);
  });
});

describe('size', () => {
  it('file size', async () => {
    expect(await snap.size('hello.txt')).toBe(toBytes('Hello!').length);
  });

  it('nested file size', async () => {
    expect(await snap.size('src/main.py')).toBe(toBytes("print('hi')").length);
  });

  it('missing throws FileNotFoundError', async () => {
    await expect(snap.size('nope.txt')).rejects.toThrow(FileNotFoundError);
  });

  it('size matches read length', async () => {
    const data = await snap.read('hello.txt');
    expect(await snap.size('hello.txt')).toBe(data.length);
  });
});

describe('objectHash', () => {
  it('returns 40-char hex string', async () => {
    const h = await snap.objectHash('hello.txt');
    expect(h).toMatch(/^[0-9a-f]{40}$/);
  });

  it('same content same hash', async () => {
    const f2 = await snap.write('copy.txt', toBytes('Hello!'));
    expect(await f2.objectHash('hello.txt')).toBe(await f2.objectHash('copy.txt'));
  });

  it('different content different hash', async () => {
    expect(await snap.objectHash('hello.txt')).not.toBe(await snap.objectHash('src/main.py'));
  });

  it('tree hash is valid hex', async () => {
    const h = await snap.objectHash('src');
    expect(h).toMatch(/^[0-9a-f]{40}$/);
  });

  it('missing throws FileNotFoundError', async () => {
    await expect(snap.objectHash('nope.txt')).rejects.toThrow(FileNotFoundError);
  });

  it('stable across calls', async () => {
    const h1 = await snap.objectHash('hello.txt');
    const h2 = await snap.objectHash('hello.txt');
    expect(h1).toBe(h2);
  });
});

describe('properties', () => {
  it('commitHash is 40-char hex', () => {
    expect(snap.commitHash).toMatch(/^[0-9a-f]{40}$/);
  });

  it('refName is main', () => {
    expect(snap.refName).toBe('main');
  });

  it('message contains last written file', async () => {
    const msg = await snap.getMessage();
    expect(msg).toContain('util.py');
  });
});

describe('readlink', () => {
  it('reads symlink target', async () => {
    const f2 = await snap.writeSymlink('link.txt', 'hello.txt');
    expect(await f2.readlink('link.txt')).toBe('hello.txt');
  });

  it('missing symlink throws FileNotFoundError', async () => {
    await expect(snap.readlink('nope.txt')).rejects.toThrow(FileNotFoundError);
  });

  it('readlink on regular file throws', async () => {
    await expect(snap.readlink('hello.txt')).rejects.toThrow(/Not a symlink/);
  });
});

// ---------------------------------------------------------------------------
// glob ** patterns on repo (ported from Python)
// ---------------------------------------------------------------------------

describe('glob', () => {
  it('** recursive pattern', async () => {
    const matches = await snap.glob('**/*.py');
    expect(matches).toContain('src/main.py');
    expect(matches).toContain('src/lib/util.py');
  });

  it('* single level', async () => {
    const matches = await snap.glob('src/*.py');
    expect(matches).toContain('src/main.py');
    expect(matches).not.toContain('src/lib/util.py');
  });

  it('no matches returns empty', async () => {
    const matches = await snap.glob('*.xyz');
    expect(matches).toEqual([]);
  });
});

// ---------------------------------------------------------------------------
// walk validates full tuple structure
// ---------------------------------------------------------------------------

describe('walk tuple structure', () => {
  it('yields [dirpath, subdirs, files] tuples', async () => {
    const entries = [];
    for await (const e of snap.walk()) entries.push(e);

    // Each entry is a 3-tuple
    for (const entry of entries) {
      expect(entry).toHaveLength(3);
      const [dirpath, subdirs, files] = entry;
      expect(typeof dirpath).toBe('string');
      expect(Array.isArray(subdirs)).toBe(true);
      expect(Array.isArray(files)).toBe(true);
    }

    // Root should have 'src' subdir and 'hello.txt' file
    const root = entries.find(([dp]) => dp === '');
    expect(root).toBeDefined();
    const [, rootDirs, rootFiles] = root!;
    expect(rootDirs).toContain('src');
    expect(rootFiles.some((f) => f.name === 'hello.txt')).toBe(true);
  });

  it('file entries have name, oid, mode', async () => {
    const entries = [];
    for await (const e of snap.walk()) entries.push(e);
    const root = entries.find(([dp]) => dp === '');
    expect(root).toBeDefined();
    const [, , files] = root!;
    const helloFile = files.find((f) => f.name === 'hello.txt');
    expect(helloFile).toBeDefined();
    expect(helloFile!.name).toBe('hello.txt');
    expect(helloFile!.oid).toMatch(/^[0-9a-f]{40}$/);
    expect(typeof helloFile!.mode).toBe('string');
  });
});

// ---------------------------------------------------------------------------
// isDir checks
// ---------------------------------------------------------------------------

describe('isDir', () => {
  it('directory returns true', async () => {
    expect(await snap.isDir('src')).toBe(true);
  });

  it('file returns false', async () => {
    expect(await snap.isDir('hello.txt')).toBe(false);
  });

  it('missing returns false', async () => {
    expect(await snap.isDir('nope')).toBe(false);
  });

  it('nested directory', async () => {
    expect(await snap.isDir('src/lib')).toBe(true);
  });
});

File: kotlin/src/test/kotlin/vost/FsReadTest.kt
package vost

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.io.FileNotFoundException
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class FsReadTest {

    @Test
    fun `read and write round-trip`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("hello.txt", "Hello, World!".toByteArray())
            val data = fs.read("hello.txt")
            assertEquals("Hello, World!", String(data))
        }
    }

    @Test
    fun `readText convenience`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("msg.txt", "Kotlin".toByteArray())
            assertEquals("Kotlin", fs.readText("msg.txt"))
        }
    }

    @Test
    fun `read nonexistent file throws`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertThrows<FileNotFoundException> {
                fs.read("nonexistent.txt")
            }
        }
    }

    @Test
    fun `read directory throws IsADirectoryError`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("dir/file.txt", "content".toByteArray())
            assertThrows<IsADirectoryError> {
                fs.read("dir")
            }
        }
    }

    @Test
    fun `read with offset and size`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("data.txt", "Hello, World!".toByteArray())
            val partial = fs.read("data.txt", offset = 7, size = 5)
            assertEquals("World", String(partial))
        }
    }

    @Test
    fun `read with offset only`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("data.txt", "Hello, World!".toByteArray())
            val data = fs.read("data.txt", offset = 7)
            assertEquals("World!", String(data))
        }
    }

    @Test
    fun `ls returns entry names`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "a".toByteArray())
            fs = fs.write("b.txt", "b".toByteArray())
            val names = fs.ls()
            assertEquals(listOf("a.txt", "b.txt"), names.sorted())
        }
    }

    @Test
    fun `ls subdirectory`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("dir/a.txt", "a".toByteArray())
            fs = fs.write("dir/b.txt", "b".toByteArray())
            val names = fs.ls("dir")
            assertEquals(listOf("a.txt", "b.txt"), names.sorted())
        }
    }

    @Test
    fun `ls on file throws`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "data".toByteArray())
            assertThrows<NotADirectoryError> {
                fs.ls("file.txt")
            }
        }
    }

    @Test
    fun `exists returns true for files and dirs`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("dir/file.txt", "data".toByteArray())
            assertTrue(fs.exists("dir/file.txt"))
            assertTrue(fs.exists("dir"))
            assertFalse(fs.exists("nonexistent"))
        }
    }

    @Test
    fun `isDir distinguishes files from dirs`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("dir/file.txt", "data".toByteArray())
            assertTrue(fs.isDir("dir"))
            assertFalse(fs.isDir("dir/file.txt"))
            assertFalse(fs.isDir("nonexistent"))
        }
    }

    @Test
    fun `fileType returns correct types`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "data".toByteArray())
            fs = fs.write("exe.sh", "#!/bin/bash".toByteArray(), mode = FileType.EXECUTABLE)
            fs = fs.writeSymlink("link", "file.txt")

            assertEquals(FileType.BLOB, fs.fileType("file.txt"))
            assertEquals(FileType.EXECUTABLE, fs.fileType("exe.sh"))
            assertEquals(FileType.LINK, fs.fileType("link"))
        }
    }

    @Test
    fun `fileType on nonexistent throws`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertThrows<FileNotFoundException> {
                fs.fileType("nonexistent")
            }
        }
    }

    @Test
    fun `size returns byte count`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            val data = "Hello, World!"
            fs = fs.write("file.txt", data.toByteArray())
            assertEquals(data.length.toLong(), fs.size("file.txt"))
        }
    }

    @Test
    fun `objectHash returns 40-char hex SHA`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "content".toByteArray())
            val hash = fs.objectHash("file.txt")
            assertEquals(40, hash.length)
            assertTrue(hash.all { c -> c in '0'..'9' || c in 'a'..'f' })
        }
    }

    @Test
    fun `commitHash returns 40-char hex SHA`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertEquals(40, fs.commitHash.length)
        }
    }

    @Test
    fun `treeHash returns 40-char hex SHA`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertEquals(40, fs.treeHash.length)
        }
    }

    @Test
    fun `refName and writable`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertEquals("main", fs.refName)
            assertTrue(fs.writable)
        }
    }

    @Test
    fun `message property`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertTrue(fs.message.contains("Initialize"))
        }
    }

    @Test
    fun `authorName and authorEmail`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertEquals("vost", fs.authorName)
            assertEquals("vost@localhost", fs.authorEmail)
        }
    }

    @Test
    fun `walk returns directory structure`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "a".toByteArray())
            fs = fs.write("dir/b.txt", "b".toByteArray())
            fs = fs.write("dir/sub/c.txt", "c".toByteArray())

            val entries = fs.walk()
            assertTrue(entries.isNotEmpty())

            // Root entry
            val root = entries[0]
            assertEquals("", root.dirpath)
            assertTrue("dir" in root.dirnames)
            assertTrue(root.files.any { it.name == "a.txt" })
        }
    }

    @Test
    fun `walk subdirectory`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("dir/a.txt", "a".toByteArray())
            fs = fs.write("dir/sub/b.txt", "b".toByteArray())

            val entries = fs.walk("dir")
            assertTrue(entries.isNotEmpty())
            assertEquals("dir", entries[0].dirpath)
        }
    }

    @Test
    fun `listdir returns WalkEntry objects`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "a".toByteArray())
            fs = fs.write("dir/b.txt", "b".toByteArray())

            val entries = fs.listdir()
            assertTrue(entries.any { it.name == "a.txt" })
            assertTrue(entries.any { it.name == "dir" && it.mode == GIT_FILEMODE_TREE })
        }
    }

    @Test
    fun `readByHash reads blob directly`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "content".toByteArray())
            val hash = fs.objectHash("file.txt")
            val data = fs.readByHash(hash)
            assertEquals("content", String(data))
        }
    }

    @Test
    fun `readByHash with offset and size`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "Hello World".toByteArray())
            val hash = fs.objectHash("file.txt")
            val data = fs.readByHash(hash, offset = 6, size = 5)
            assertEquals("World", String(data))
        }
    }

    @Test
    fun `readlink returns symlink target`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.writeSymlink("link", "target.txt")
            assertEquals("target.txt", fs.readlink("link"))
        }
    }

    @Test
    fun `readlink on non-symlink throws`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "data".toByteArray())
            assertThrows<IllegalStateException> {
                fs.readlink("file.txt")
            }
        }
    }

    @Test
    fun `write returns new Fs with different commit`() {
        val store = createStore()
        store.use {
            val fs1 = it.branches["main"]
            val fs2 = fs1.write("file.txt", "data".toByteArray())
            assertTrue(fs1.commitHash != fs2.commitHash)
            assertEquals("data", String(fs2.read("file.txt")))
        }
    }

    @Test
    fun `nested directory creation`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a/b/c/d.txt", "deep".toByteArray())
            assertEquals("deep", fs.readText("a/b/c/d.txt"))
            assertTrue(fs.isDir("a"))
            assertTrue(fs.isDir("a/b"))
            assertTrue(fs.isDir("a/b/c"))
        }
    }

    @Test
    fun `empty initial repo ls returns empty`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertEquals(emptyList(), fs.ls())
        }
    }

    @Test
    fun `toString includes commit hash prefix`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val str = fs.toString()
            assertTrue(str.startsWith("Fs("))
            assertTrue("main" in str)
        }
    }

    // ── Root-path handling ──────────────────────────────────────────────

    @Test
    fun `exists returns true for root paths`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertTrue(fs.exists(""))
            assertTrue(fs.exists("/"))
        }
    }

    @Test
    fun `isDir returns true for root`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertTrue(fs.isDir(""))
        }
    }

    @Test
    fun `fileType returns TREE for root`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertEquals(FileType.TREE, fs.fileType(""))
        }
    }

    @Test
    fun `size on root throws IsADirectoryError`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertThrows<IsADirectoryError> {
                fs.size("")
            }
        }
    }

    @Test
    fun `objectHash on root returns treeHash`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            assertEquals(fs.treeHash, fs.objectHash(""))
        }
    }

    // ── Ranged read overflow ────────────────────────────────────────────

    @Test
    fun `read with large offset and size does not overflow`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "hello".toByteArray())
            val data = fs.read("file.txt", offset = Int.MAX_VALUE - 10, size = 100)
            assertEquals(0, data.size)
        }
    }

    @Test
    fun `readByHash with large offset and size does not overflow`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "hello".toByteArray())
            val hash = fs.objectHash("file.txt")
            val data = fs.readByHash(hash, offset = Int.MAX_VALUE - 10, size = 100)
            assertEquals(0, data.size)
        }
    }
}
